cwms-data-api/src/main/java/cwms/cda/data/dto/location/kind/PhysicalStructureChange.java

```diff
@@ -0,0 +1,250 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package cwms.cda.data.dto.location.kind;
+
+import static java.util.Comparator.comparing;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import cwms.cda.data.dto.CwmsDTO;
+import cwms.cda.data.dto.CwmsDTOValidator;
+import cwms.cda.data.dto.CwmsId;
+import cwms.cda.data.dto.LookupType;
+import cwms.cda.formatters.Formats;
+import cwms.cda.formatters.annotations.FormattableWith;
+import cwms.cda.formatters.json.JsonV1;
+import java.time.Instant;
+import java.util.Set;
+import java.util.TreeSet;
+
+@FormattableWith(contentType = Formats.JSONV1, formatter = JsonV1.class, aliases = {Formats.DEFAULT, Formats.JSON})
+@JsonDeserialize(builder = PhysicalStructureChange.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+@JsonPropertyOrder({"project-id", "change-date", "protected", "discharge-computation-type", "reason-type", "notes"})
+public final class PhysicalStructureChange<T extends Setting> extends CwmsDTO {
+    @JsonProperty(required = true)
+    private final CwmsId projectId;
+    @JsonProperty(value = "protected", required = true)
+    private final boolean isProtected;
+    @JsonProperty(required = true)
+    private final Instant changeDate;
+    @JsonProperty(required = true)
+    private final LookupType dischargeComputationType;
+    @JsonProperty(required = true)
+    private final LookupType reasonType;
+    private final Double newTotalDischargeOverride;
+    private final Double oldTotalDischargeOverride;
+    private final String dischargeUnits;
+    private final Double poolElevation;
+    private final Double tailwaterElevation;
+    private final String elevationUnits;
+    private final String notes;
+    private final Set<T> settings;
+
+    private PhysicalStructureChange(Builder<T> builder) {
+        //uses CwmsId instead of a separate office-id field
+        super(null);
```

Comment:
A thought, would it make sense to put a CwmsId at the CwmsDto level, and use a method over override the output name? Though we'd have to put an @JsonIgnore at the CwmsDto level.

Just doesn't seem nice to have a special case that ignores that root element.

Reply:
yeah this definitely seems weird to me, but when I tried putting CwmsId in CwmsDto that also got weird. We don't have any need for the cwmsId except for getting the office id out of it and it doesn't solve the need for @JsonIgnore on the "office-id" fields. We do already have precedence for handing null to the super in several classes so that's why I went with this solution.

A slightly cleaner idea might be to create a new abstract base class and move the getOfficeId() method up to the CwmsDTOBase interface and then mark the abstract class's implementation of getOfficeId() and the cwmsId field as @JsonIgnore'd.

The other issue is that in classes like PhysicalStructure, we have a CwmsId for the project-id, but really the identifying information for the DTO is the Location which isn't and doesn't have a CwmsId (might be worth having in a v3 of the DTO?).

Reply:
Yeah. I don't see a good solution right now. as long as a getOffice method returns something reasonable though that's the important part. So maybe you're right, CwmsDto should just be an interface with getOffice (or maybe specifically getOfficeForWritePermissions) to be able to set the connection and the objects do what they need. Something to consider later though.